### PR TITLE
Feature/proj 145

### PR DESCRIPTION
**Pull request by Lelik7777**
# Profile API Integration and Data Management

### 1. Profile API and Types Update
- Added constants for profile API endpoints in `config/api.ts`
- Updated `baseApi.ts` to include new API endpoints for profile operations
- Enhanced `profile.types.ts`:
  - Added JSDoc comments for better code documentation
  - Updated types and introduced interfaces for improved type safety
  - Restructured types to better represent the API response and request structures

### 2. User Data and Avatar Upload Implementation
- Implemented functionality to upload user profile data to the server
- Added avatar upload capability, including handling of both JPEG and PNG formats
- Created `imageUtils.ts` for converting image Data URLs to a format suitable for server upload

### 3. Code Refactoring and Optimization
- Refactored `useGeneralInfo` hook to use the new API endpoints and types
- Optimized image processing logic for better performance and reusability

### Profile API Integration
- Added new endpoints in `baseApi.ts` for operations like:
  - Fetching user profile
  - Updating user profile
  - Uploading avatar
  - Deleting avatar and profile

### Types and Interfaces Update
- Updated `ProfileResponse` and `ProfileSettings` interfaces in `profile.types.ts`
- Added comprehensive JSDoc comments to explain the purpose and structure of each type/interface

### Image Upload Utility
- Created `imageUtils.ts` with a `prepareImageForUpload` function to handle image data conversion
- This utility supports both JPEG and PNG formats and prepares the image data for server upload

### Profile Update Logic
- Enhanced `useGeneralInfo` hook to handle both profile data and avatar updates
- Implemented error handling and success notifications for profile update operations

## Testing Instructions
1. Test profile data fetching and display
2. Attempt to update profile information and verify the changes are reflected
3. Upload different image formats (JPEG, PNG) as avatars and ensure they are processed correctly
4. Verify error handling by intentionally causing errors (e.g., network disconnection during upload)

## Notes
- Ensure the backend API endpoints match the new constants defined in the frontend
- The image upload utility may need adjustments based on specific backend requirements

Please review and test thoroughly, focusing on the integration points between the new API calls and the existing UI components.

### CODE REVIEW

File: src/features/profile/api/profile.types.ts [hunk 1]
```diff
@@ -1,28 +1,19 @@
...
-  lastName: string
-  region: string
-  userName: string
-}
-
-export type ProfileSettings = {
+/**
+ * base profile interface
+ */
+interface BaseProfile {
```
Comment (EvgeniyaYakunina) [inline src/features/profile/api/profile.types.ts:4]: Зачем interface? Мы везде type используем.

File: src/features/profile/api/profile.types.ts [hunk 2]
```diff
@@ -32,17 +23,69 @@ export type Avatar = {
...
+ *       "height": 200,
+ *       "fileSize": 15000,
+ *       "createdAt": "2024-09-23T10:30:00Z"
+ *     }
+ *   ],
+ *   "createdAt": "2024-01-01T00:00:00Z",
+ *   "dateOfBirth": "1990-01-01"
+ * }
+ */
+export interface ProfileResponse extends BaseProfile {
```
Comment (EvgeniyaYakunina) [inline src/features/profile/api/profile.types.ts:50]: тут тоже type просто можно использовать.

File: src/features/profile/api/profile.types.ts [hunk 3]
```diff
@@ -32,17 +23,69 @@ export type Avatar = {
...
+ *   country: "USA",
+ *   firstName: "Jane",
+ *   lastName: "Doe",
+ *   region: "California",
+ *   userName: "janedoe_dev"
+ * }
+ */
+export type ProfileSettings = {
+  dateOfBirth: string
+} & BaseProfile
```
Comment (EvgeniyaYakunina) [inline src/features/profile/api/profile.types.ts:73]: Какой-то странный тип. У тебя же дата рождения уже есть в типе ProfileResponse

File: src/features/profile/api/profileApi.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
+import { baseApi } from '@/app/api'
+import { API_URLS, POST } from '@/shared/config'
+
+import { ProfileResponse, ProfileSettings, SuccessAvatarResponse } from './profile.types'
+
+export const profileApi = baseApi.injectEndpoints({
+  endpoints: builder => ({
+    /**
+     * Delete the user's avatar.
+     * @returns {Promise<void>}
```
Comment (EvgeniyaYakunina) [inline src/features/profile/api/profileApi.ts]: мне кажется это уточнение лишнее.Тут это и так понятно.

File: src/features/profile/api/profileApi.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    deleteAvatar: builder.mutation<void, void>({
+      invalidatesTags: ['Profile'],
+      query: () => ({
+        method: POST,
+        url: API_URLS.PROFILE.DELETE_AVATAR,
+      }),
+    }),
+    /**
+     * Delete the user's profile.
+     * @returns {Promise<void>}
```
Comment (EvgeniyaYakunina) [inline src/features/profile/api/profileApi.ts]: тоже лишнее

File: src/features/profile/api/profileApi.ts [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+      invalidatesTags: ['Profile'],
+      query: () => ({
+        method: POST,
+        url: API_URLS.PROFILE.DELETE_PROFILE,
+      }),
+    }),
+    /**
+     * Delete a profile by user ID.
+     * @param {number} userId - The ID of the user whose profile is to be deleted.
+     * @returns {Promise<void>}
```
Comment (EvgeniyaYakunina) [inline src/features/profile/api/profileApi.ts]: давай нигде не будем писать, если возвращается void.Мне кажется это очевидная вещь.

File: src/app/api/baseQueryWithReauth.ts [hunk 7]
```diff
@@ -35,7 +35,7 @@ export const baseQueryWithReauth: BaseQueryFn<
         const refreshResult = await baseQuery(
           {
             credentials: 'include',
-            method: POST,
+            method: METHOD.POST,
```
Comment (EvgeniyaYakunina) [inline src/app/api/baseQueryWithReauth.ts]: А зачем эта константа METHOD? мне кажется это уже масло масляное.

File: src/shared/config/apiURLs.ts [hunk 8]
```diff
@@ -23,4 +23,9 @@ export const API_URLS = {
   },
 }
 
-export const POST = 'POST'
+export const METHOD = {
+  DELETE: 'DELETE',
+  GET: 'GET',
+  POST: 'POST',
+  PUT: 'PUT',
+} as const
```
Comment (EvgeniyaYakunina) [inline src/shared/config/apiURLs.ts:31]: Это извращенство) Давай не будем все подряд в константы превращать, это уже перебор.

File: src/features/profile/api/profileApi.ts (hunk 4 above)
Comment (Lelik7777) [reply src/features/profile/api/profileApi.ts]: исправил

File: src/features/profile/api/profileApi.ts (hunk 5 above)
Comment (Lelik7777) [reply src/features/profile/api/profileApi.ts]: исправил

File: src/features/profile/api/profile.types.ts (hunk 1 above)
Comment (Lelik7777) [reply src/features/profile/api/profile.types.ts:4]: для наследования

File: src/app/api/baseQueryWithReauth.ts (hunk 7 above)
Comment (Lelik7777) [reply src/app/api/baseQueryWithReauth.ts]: исправил

**Review by EvgeniyaYakunina — APPROVED** [review]

**Comment by Lelik7777** [comment]
run build production

**Pull request merged**